### DISCUSSION TITLE
Using a custom select object

### DISCUSSION THREAD

**Discussion by ShmuelTreiger** [sqlalchemy/sqlalchemy#6693] [Usage Questions]
Now that session.query isn't recommended, is there a way to create a custom Select object? Similar to this stackoverflow question: https://stackoverflow.com/questions/15936111/sqlalchemy-can-you-add-custom-methods-to-the-query-object

I've looked around and haven't found anything, just thought I would see if I'm missing something before raising an issue.

So far I have been able to create the custom class and inherit from select, but I have been unable to add it to the session.

```
class CustomSelect(Select):
    def custom_method(self):
        pass

Session(select_cls=CustomSelect, future=True)

```
The class works, the session doesn't

**Comment by CaselIT** [comment]
Hi,

I'm not sure what you mean by "adding it to the session"
A select is not created by a session, but by using the `select()` constructor, then it's executed by a session using `session.execute(session_obj)`

In your case you can simply instanciate your select and execute it as usual.

What's your use case for needing to subclass a select?

**Reply by ShmuelTreiger** [reply]
OK, this makes sense to me. Whereas I used to do:

`session.query()`

instead I'm doing

`session.execute(select())`

However, I'm still unclear as to how I can augment the select. For example, I'd like to do:

`statement = select(Table).filter(Table.id == 1).special_custom_filter()`

Is this possible?

Thinking about it, `select()` returns a Select object, which has a .filter() method. So I'd like `select()` to return a CustomSelect object which extends Select and implements my methods. Not sure how to do it, but I don't think it's an SQLAlchemy problem.

Thanks!

**Reply by CaselIT** [reply]
`select()` is a constructor for `Select`, that just calls `_create`:
https://github.com/sqlalchemy/sqlalchemy/blob/67edf8e376d4bf4baf5944b88dff1cb073e9ef72/lib/sqlalchemy/sql/selectable.py#L4971

regarding creating a subclass it should be possible to do it with something like:
```py
class MySelect(Select):
  @classmethod
  def create(cls, *args, **kw):
    return cls._create(*args, **kw)

  def my_method(self, ...):
    pass

my_select = MySelect.create(...)
```

@zzzeek do you have a better solution?

**Reply by zzzeek** [reply]
the current organization of `Select` seems to be lacking a straightforward way to make a subclass without using a private method for construction, this is largely a product of the 1.4/2.0 thing where we are trying to limit the API to clear-cut "future" and "legacy" patterns.  I think we should add a public helper method and an example.  however, for now this is how to make a custom select() subclass:

```py
from sqlalchemy import select, column
from sqlalchemy.sql.expression import Select


class MySelect(Select):
    def my_filter(self, value):
        return self.filter(column("q") == value)


def my_select(*expr):
    return MySelect._create_future_select(*expr)


s1 = (
    my_select(column("p"))
    .where(column("z") == 10)
    .my_filter(12)
    .order_by(column("y"))
    .my_filter(9)
)

print(s1)
```

**Reply by ShmuelTreiger** [reply]
Awesome. Thank you both so much! I'll give this a go at some point today/tonight

If/when you create the issue/patch, could you please tag me so I can get updated on this feature? Super appreciate it!

**Reply by CaselIT** [reply]
Added  #6700 to track this

**Reply by fostermaier** [reply]
Thank you all so much!

I was failing by implementing a custom `__new__()` method when i stumbled upon this disussion. Renaming the class method to something else and returning a `cls._create_future_select()` object works like a charm.
